pyoptsparse/meson.build

```diff
@@ -0,0 +1,104 @@
+# NumPy include directory - needed in all submodules
+incdir_numpy = get_option('incdir_numpy')
+if incdir_numpy == ''
+    incdir_numpy = run_command(py3_target,
+      [
+        '-c',
+        'import os; os.chdir(".."); import numpy; print(numpy.get_include())'
+      ],
+      check: true
+    ).stdout().strip()
+endif
+# this creates a raw string which is useful for Windows use of '\' for paths
+incdir_numpy = '''@0@'''.format(incdir_numpy)
+
+# HACK: Meson prefixes filenames of intermediate compiled objects with their filepath. This poses a problem for conda builds
+# since conda ensures the host environment directory has 255 characters so the meson object filenames then exceed 255
+# characters. To remedy this, the fortranobject.c file from numpy is copied into pyoptsparse so that the meson build
+# uses a relative path, rather than an absolute path, thus reducing the auto generated object filename
+# see for example https://github.com/mesonbuild/meson/issues/4226
+run_command(py3_command,
+  [
+    '-c',
+    'import os; os.chdir(".."); import shutil; shutil.copy(os.path.join(r"' + incdir_numpy + '", "..", "..", "f2py", "src", "fortranobject.c"), "pyoptsparse")'
+  ],
+  check: true
+)
+
+inc_np = include_directories(incdir_numpy)
+
+
+# TODO: pyoptsparse supports numpy>=1.16 but numpy.f2py.get_include() wasnt added until later, raise numpy version?
+#incdir_f2py = run_command(py3_target,
+#  [
+#    '-c',
+#    'import os; os.chdir(".."); import numpy.f2py; print(numpy.f2py.get_include())'
+#  ],
+#  check : true
+#).stdout().strip()
+incdir_f2py = incdir_numpy / '..' / '..' / 'f2py' / 'src'
+inc_f2py = include_directories(incdir_f2py)
+
+# Don't use the deprecated NumPy C API. Define this to a fixed version instead of
+# NPY_API_VERSION in order not to break compilation for released SciPy versions
+# when NumPy introduces a new deprecation. Use in a meson.build file::
+#
+#   py3_target.extension_module('_name',
+#     'source_fname',
+#     numpy_nodepr_api)
+#
+# numpy_nodepr_api = '-DNPY_NO_DEPRECATED_API=NPY_1_9_API_VERSION'
+
+#python_sources = [
+#    '__init__.py',
+#    'pyOpt_MPI.py',
+#    'pyOpt_constraint.py',
+#    'pyOpt_error.py',
+#    'pyOpt_gradient.py',
+#    'pyOpt_history.py',
+#    'pyOpt_objective.py',
+#    'pyOpt_optimization.py',
+#    'pyOpt_optimizer.py',
+#    'pyOpt_solution.py',
+#    'pyOpt_utils.py',
+#    'pyOpt_variable.py',
+#    'types.py'
+#]
+
+#py3_target.install_sources(
+#  python_sources,
+#  pure: true,
+#  subdir: 'pyoptsparse'
+#)
```

Comment:
These are kept because in the future, meson may become pep517 compliant, which means we don't need to invoke pip at all. In such cases, meson would need to know which source files (e.g. all python files) to install for a user, that's what these lines do.

Reply:
Gotcha, I did not realize that the pep517 issue involved these as well. I would just add a `#TODO` in case it's not there already